### PR TITLE
feat(op): Cancun support in payload builder and processor

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Implements Cancun support in the `optimism` payload builder and the EVM processor.

**closes**
https://github.com/paradigmxyz/reth/issues/4213

### CODE REVIEW

File: crates/primitives/src/revm/env.rs [hunk 1]
```diff
@@ -156,7 +156,9 @@ pub fn fill_tx_env_with_beacon_root_contract_call(env: &mut Env, parent_beacon_b
             source_hash: None,
             mint: None,
             is_system_transaction: Some(false),
-            enveloped_tx: None,
+            // The L1 fee is not charged for the EIP-4788 transaction, submit zero bytes for the
+            // enveloped tx size.
+            enveloped_tx: Some(Bytes::default()),
         },
```
Comment (gakonst) [inline crates/primitives/src/revm/env.rs:162]: @mattsse @Rjected confirming this is OK?

Comment (Rjected) [reply crates/primitives/src/revm/env.rs:162]: yeah this should be fine, i think we eventually want to deprecate the `enveloped_tx` field and replace it with
```rust
struct L1TxFeeInfo {
    zeros: u32,
    ones: u32
}
```
or something

File: crates/payload/builder/src/payload.rs [hunk 2]
```diff
@@ -73,9 +73,34 @@ impl EthBuiltPayload {
...
+            // heuristic this flag **SHOULD** be set to `false`.
+            //
+            // Spec:
+            // <https://github.com/ethereum/execution-apis/blob/fe8e13c288c592ec154ce25c534e26cb7ce0530d/src/engine/cancun.md#specification-2>
+            should_override_builder: false,
+            blobs_bundle: sidecars.into_iter().map(Into::into).collect::<Vec<_>>().into(),
+            // Optimism-specific: Post-cancun, the parent beacon block root is included in the
+            // enveloped payload.
+            parent_beacon_block_root,
+        }
```
Comment (Rjected) [inline crates/payload/builder/src/payload.rs:103]: is there a way to not do this in main eth conversions?

File: crates/net/network/src/transactions/fetcher.rs
Comment (mattsse) [inline crates/net/network/src/transactions/fetcher.rs:1]: please undo

**Review by mattsse — CHANGES_REQUESTED** [review]
this modifies the ethbuiltpayload to incorporate checks that we need for OP

not a fan of this, this needs a OP specific solution.

we should change this instead:

https://github.com/paradigmxyz/reth/blob/761c707e418bbe90bd2e47357ecf1673185f6c2a/crates/node-optimism/src/engine.rs#L17-L17

File: crates/net/network/src/transactions/fetcher.rs
Comment (clabby) [reply crates/net/network/src/transactions/fetcher.rs:1]: Do we want `;`'s now? `rustfmt.toml` still has `trailing_semicolon = false`.

File: crates/payload/builder/src/payload.rs [hunk 3]
```diff
@@ -73,9 +73,34 @@ impl EthBuiltPayload {
         self.into()
     }
 
-    /// Converts the type into the response expected by `engine_getPayloadV2`
-    pub fn into_v3_payload(self) -> ExecutionPayloadEnvelopeV3 {
-        self.into()
+    /// Converts the type into the response expected by `engine_getPayloadV3`
+    pub fn into_v3_payload(self, spec: &ChainSpec) -> ExecutionPayloadEnvelopeV3 {
```
[the remark this answers is not in the document]
Comment (clabby) [reply crates/payload/builder/src/payload.rs]: wdyt about https://github.com/paradigmxyz/reth/pull/6731/commits/d3ff7edd8b0b710927ad1767e161a833d402a14b

File: crates/rpc/rpc-engine-api/src/engine_api.rs [hunk 4]
```diff
@@ -265,13 +265,24 @@ where
...
+        let mut resolved_payload = self
             .inner
             .payload_store
             .resolve(payload_id)
             .await
             .ok_or(EngineApiError::UnknownPayload)?
-            .map(|payload| payload.into_v3_payload())?)
+            .map(|payload| payload.into_v3_payload())?;
+
+        // After `Cancun` is enabled on optimism, an extra field `parent_beacon_block_root` is
```
Comment (mattsse) [inline crates/rpc/rpc-engine-api/src/engine_api.rs:276]: I'm okay with this for now, but ideally this should be solved with the Op payload type instead

File: crates/rpc/rpc-engine-api/src/engine_api.rs [hunk 5]
```diff
@@ -265,13 +265,24 @@ where
...
+        // included in the enveloped V3 payload. On ethereum, this field is not included.
+        resolved_payload.parent_beacon_block_root = (self.inner.chain_spec.is_optimism() &&
```
Comment (mattsse) [inline crates/rpc/rpc-engine-api/src/engine_api.rs:278]: please move this update in an if

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/network/src/transactions/fetcher.rs
Comment (mattsse) [inline crates/net/network/src/transactions/fetcher.rs:1]: please undo

File: crates/transaction-pool/src/pool/best.rs
Comment (mattsse) [inline crates/transaction-pool/src/pool/best.rs:1]: please undo

File: crates/transaction-pool/src/blobstore/disk.rs
Comment (mattsse) [inline crates/transaction-pool/src/blobstore/disk.rs:1]: please undo

**Review by mattsse — CHANGES_REQUESTED** [review]

**Review by mattsse — APPROVED** [review]
we're back 🫡
ty

**Pull request merged**